File: src/schemas/schemas.py
# Market & Options Data Contracts
# Shared contract between the TypeScript SDK and AI agent

from pydantic import BaseModel
from typing import Optional, Literal


"""
Greeks associated with an option
"""

class OptionGreeks(BaseModel):
    delta: float
    iv: float
    gamma: float
    theta: float
    vega: float


class DemoFillPreview(BaseModel):
    fillSizeUsdc: Optional[float | int | str] = None
    numContracts: Optional[float | int | str] = None
    totalCollateral: Optional[float | int | str] = None


"""
A single option order retrived from Thetanuts Optionbook

This is the normalized representation used by the application
"""

class OptionOrder(BaseModel):
    maker: str
    orderExpiryTimestamp: int
    expiry: int
    collateral: str
    isCall: bool
    priceFeed: str
    implementation: str
    isLong: bool
    maxCollateralUsable: str
    strikes: list[int]
    price: str
    numContracts: str
    extraOptionData: str
    signature: str
    nonce: str
    greeks: Optional[OptionGreeks] = None
    demoFillPreview: Optional[DemoFillPreview] = None


"""
The Available options from Thetanuts Optionbook.
"""

class OptionBookData(BaseModel):
    chainId: int
    optionBookAddress: str
    orders: list[OptionOrder]
    lastUpdated: int
    currentTime: int


"""
Current underlying market condition
"""

class VolatilityData(BaseModel):
    current: float
    forecast: list[float]

class MarketState(BaseModel):
    prices: dict[str, float]
    volatility: dict[str, VolatilityData]


"""
Combined market information given to the AI agent
"""
class TradingMarketData(BaseModel):
    market: Optional[MarketState] = None
    optionBook: Optional[OptionBookData] = None
    retrievedAt: int

"""
Request made by AI agent when it needs market data
"""

class MarketDataRequest(BaseModel):
    asset: str
    includeOptions: bool
    includeMarketState: bool
    optionType: Optional[Literal["call", "put", "both"]] = None
    side: Optional[Literal["long", "short", "both"]] = None
    expiryAfter: Optional[int] = None
    expiryBefore: Optional[int] = None
    strikeMin: Optional[int] = None
    strikeMax: Optional[int] = None
    requireGreeks: Optional[bool] = None
    limit: Optional[int] = None

File: src/Schemas/schemas.ts
// Market & Options Data Contracts
// Shared contract between the TypeScript SDK and AI agent

/*
Greeks associated with an option
*/

export interface OptionGreeks {
    delta: number;
    iv: number;
    gamma: number;
    theta: number;
    vega: number;
}

/*
A single option order retrived from Thetanuts Optionbook

This is the normalized representation used by the application
*/

export interface OptionOrder {
    maker: string;
    orderExpiryTimestamp: number;
    expiry: number;
    collateral: string;
    isCall: boolean;
    priceFeed: string;
    implementation: string;
    isLong: boolean;
    maxCollateralUsable: string;
    strikes: number[];
    price: string;
    numContracts: string;
    extraOptionData: string;
    signature: string;
    nonce: string;
    greeks?: OptionGreeks;
}

/*
The Available options from Thetanuts Optionbook.
*/

export interface OptionBookData {
    chainId: number;
    optionBookAddress: string;
    orders: OptionOrder[];
    lastUpdated: number;
    currentTime: number;
}

/*
Current underlying market condition
*/

export interface MarketState {
    prices: Record<string, number>;
    volatility: Record<string, 
    {
        current: number;
        forecast: number[];
    }
    >;
}

/*
Combined market information given to the AI agent
*/

export interface TradingMarketData {
    market: MarketState;
    optionBook: OptionBookData;
    retrievedAt: number;
}

/* 
Request made by AI agent when it needs market data
*/

export interface MarketDataRequest {
    asset: string;
    includeOptions: boolean;
    includeMarket: boolean;
    optionType?: "call" | "put" | "both";
    side?: "long" | "short" | "both";
    expiryAfter?: number;
    expiryBefore?: number;
    strikeMin?: number;
    strikeMax?: number;
    limit?: number;
}